packages/components/src/DateTimePickers/InputDateTimePicker/InputDateTimePicker.component.js

```diff
@@ -100,31 +104,30 @@ class InputDateTimePicker extends React.Component {
 		}
 	}
 
-	onChange(event, nextState, origin) {
-		const { errorMessage, datetime, textInput } = nextState;
-
-		const datetimeUpdated =
-			datetime !== this.state.datetime && !isSameSecond(datetime, this.state.datetime);
+	onChange(event, origin) {
+		const { errorMessage, datetime, textInput } = this.state;
 
 		const errorUpdated = errorMessage !== this.state.errorMessage;
 
-		this.setState(nextState, () => {
-			if (this.props.onChange && (datetimeUpdated || errorUpdated)) {
-				this.props.onChange(event, { errorMessage, datetime, textInput, origin });
-			}
-		});
+		// updating initial state
+		this.initialState = { ...this.state };
+		if (this.props.onChange && (this.state.datetimeUpdated || errorUpdated)) {
+			this.props.onChange(event, { errorMessage, datetime, textInput, origin });
+		}
 	}
 
 	onInputChange(event) {
 		const textInput = event.target.value;
 		const nextState = extractPartsFromTextInput(textInput, this.getDateOptions());
-		return this.onChange(event, nextState, 'INPUT');
+		const datetimeUpdated = this.dateHasChanged(nextState);
```

Comment:
Can you remove all those computations everytime there is a change ?
You have everything you need in this.state and this.initialState, you can compute it in `this.onChange()` to see if you trigger the onChange callback.
It's less error prone to have it in 1 place